src/ccall.cpp

```diff
@@ -370,62 +250,71 @@ static Value *julia_to_native(Type *ty, jl_value_t *jt, Value *jv,
             }
         }
-        // error. box for error handling.
-        jv = boxed(jv,ctx);
+        emit_error("ccall: argument type did not match declaration", ctx);
+    }
+    if (jl_is_tuple(jt)) {
+        return emit_unbox(ty,jv,jt);
     }
-    else if (jl_is_cpointer_type(jt)) {
+    if (jl_is_cpointer_type(jt) && addressOf) {
         assert(ty->isPointerTy());
+        jl_value_t *ety = jl_tparam0(jt);
         jl_value_t *aty = expr_type(argex, ctx);
-        if (jl_is_array_type(aty) &&
-            (jl_tparam0(jt) == jl_tparam0(aty) || jl_tparam0(jt) == (jl_value_t*)jl_bottom_type ||
-             jl_tparam0(jt) == (jl_value_t*)jl_void_type)) {
-            // array to pointer
-            return builder.CreateBitCast(emit_arrayptr(jv), ty);
-        }
-        if (aty == (jl_value_t*)jl_ascii_string_type || aty == (jl_value_t*)jl_utf8_string_type) {
-            return builder.CreateBitCast(emit_arrayptr(emit_nthptr(jv,1,tbaa_const)), ty);
+        if (aty != ety && ety != (jl_value_t*)jl_any_type && jt != (jl_value_t*)jl_voidpointer_type) {
+            std::stringstream msg;
+            msg << "ccall argument ";
+            msg << argn;
+            emit_typecheck(jv, ety, msg.str(), ctx);
         }
-        if (jl_is_structtype(aty) && jl_is_leaf_type(aty) && !jl_is_array_type(aty)) {
-            if (!addressOf) {
-                emit_error("ccall: expected & on argument", ctx);
-                return literal_pointer_val(jl_nothing);
-            }
+        if (jl_is_mutable_datatype(ety)) {
+            // no copy, just reference the data field
             return builder.CreateBitCast(emit_nthptr_addr(jv, (size_t)1), ty); // skip type tag field
         }
-        *mightNeedTempSpace = true;
-        Value *p = builder.CreateCall4(prepare_call(value_to_pointer_func),
-                                       literal_pointer_val(jl_tparam0(jt)), jv,
-                                       ConstantInt::get(T_int32, argn),
-                                       ConstantInt::get(T_int32, (int)addressOf));
+        else if (jl_is_immutable_datatype(ety)) {
+            // yes copy
+            Value *nbytes;
+            if (jl_is_leaf_type(ety))
+                nbytes = ConstantInt::get(T_int32, jl_datatype_size(ety));
+            else
+                nbytes = emit_nthptr_recast(emit_typeof(jv), offsetof(jl_datatype_t,size)/sizeof(char*), tbaa_datatype, T_int32);
+            *needStackRestore = true;
+            AllocaInst *ai = builder.CreateAlloca(T_int8, nbytes);
+            ai->setAlignment(128); //bits = 16 bytes
+            builder.CreateMemCpy(ai, builder.CreateBitCast(emit_nthptr_addr(jv, (size_t)1), T_pint8), nbytes, 1);
+            return builder.CreateBitCast(ai, ty);
+        }
+        // emit maybe copy
+        *needStackRestore = true;
+        Value *jvt = emit_typeof(jv);
+        Value *p;
+        BasicBlock *mutableBB = BasicBlock::Create(getGlobalContext(),"is-mutable",ctx->f);
```

Comment:
(added WIP title)

yeah, i'll definately have to do more testing on this. i probably will hold off on merging until I get a few more of the ccall-related PR's ready. the runtime code branch is identical to the compile-time code branch directly above it, however.